numpy/core/_internal.py

```diff
@@ -355,15 +355,38 @@ def _view_is_safe(oldtype, newtype):
         If the new type is incompatible with the old type.
 
     """
+    from numpy.lib.type_check import find_dtype_offsets
 
     # if the types are equivalent, there is no problem.
     # for example: dtype((np.record, 'i4,i4')) == dtype((np.void, 'i4,i4'))
     if oldtype == newtype:
         return
 
-    if newtype.hasobject or oldtype.hasobject:
-        raise TypeError("Cannot change data-type for object array.")
-    return
+    # no object members is fine
+    if not newtype.hasobject and not oldtype.hasobject:
+        return
```

Comment:
Perhaps the best solution is just to store a reference to the original dtype for partial views, or simply a list of hidden object offsets?